walletnode/services/userdataprocess/task.go

```diff
@@ -0,0 +1,391 @@
+package userdataprocess
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/pastelnetwork/gonode/common/errgroup"
+	"github.com/pastelnetwork/gonode/common/errors"
+	"github.com/pastelnetwork/gonode/common/log"
+	"github.com/pastelnetwork/gonode/common/service/task"
+	"github.com/pastelnetwork/gonode/common/service/userdata"
+	"github.com/pastelnetwork/gonode/walletnode/services/userdataprocess/node"
+)
+
+// Task is the task of userdata processing.
+type Task struct {
+	task.Task
+	*Service
+
+	// information of nodes process to set userdata
+	nodes      node.List
+	resultChan chan *userdata.ProcessResult
+	err        error
+	request    *userdata.ProcessRequest
+
+	// information user pastelid to retrieve userdata
+	userpastelid  string // user pastelid
+	resultChanGet chan *userdata.ProcessRequest
+}
+
+// Run starts the task
+func (task *Task) Run(ctx context.Context) error {
+	ctx = log.ContextWithPrefix(ctx, fmt.Sprintf("%s-%s", logPrefix, task.ID()))
+
+	log.WithContext(ctx).Debugf("Start task")
+	defer log.WithContext(ctx).Debugf("End task")
+	defer close(task.resultChan)
+	defer close(task.resultChanGet)
+
+	if err := task.run(ctx); err != nil {
+		task.err = err
+		task.UpdateStatus(StatusTaskFailure)
+		log.WithContext(ctx).WithError(err).Warnf("Task failed")
+
+		return nil
+	}
+
+	task.UpdateStatus(StatusTaskCompleted)
+
+	return nil
+}
+
+func (task *Task) run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	maxNode := task.config.NumberSuperNodes
+	if task.request == nil {
+		// This is process to retrieve userdata
+		maxNode = 1 // Get data from 1 supernode only, currently we choose the 1st ranked supernode, but this may change later
+	}
+
+	// TODO: Make this init and connect to super nodes to generic reusable function to avoid code duplication (1)
+	// Retrieve supernodes with highest ranks.
+	topNodes, err := task.pastelTopNodes(ctx, maxNode)
+	if err != nil {
+		return err
+	}
+	if len(topNodes) < maxNode {
+		task.UpdateStatus(StatusErrorNotEnoughMasterNode)
+		return errors.New("unable to find enough Supernodes to send userdata to")
+	}
+
+	// Try to create mesh of supernodes, connecting to all supernodes in a different sequences.
+	var nodes node.List
+	var errs error
+	nodes, err = task.meshNodes(ctx, topNodes, 0) // Connect a mesh node with primary is 1st ranked SN
```

Comment:
The node at index 0 doesn't always available. Should try from 0 -> maxNode and break at the first one successful.

Reply:
Actually the requirement require when we run in production, 10 supernode will be used to verify the data. So right now this isn't need, because we need 10 SN and every of them must be available. But I'll take note and fix it in the future if number of require supernode to verify the data is reduced, or masternode top return more than 10 SNs.